server/src/main/java/io/crate/expression/scalar/string/TranslateFunction.java

```diff
@@ -0,0 +1,144 @@
+import io.crate.types.DataTypes;
+import io.crate.types.TypeSignature;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class TranslateFunction extends Scalar<String, String> {
+
+    private static final TypeSignature STRING = DataTypes.STRING.getTypeSignature();
+    private static final char NULL = '\0';
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(
+            Signature.scalar("translate", STRING, STRING, STRING, STRING),
+            TranslateFunction::new
+        );
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+    private final Map<Character, Character> tmap;
+
+    public TranslateFunction(Signature signature, Signature boundSignature) {
+        this(signature, boundSignature, null);
+    }
+
+    public TranslateFunction(Signature signature, Signature boundSignature, Map<Character, Character> tmap) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+        this.tmap = tmap;
+    }
+
+    @Override
+    public Scalar<String, String> compile(List<Symbol> args) {
+        assert args.size() == 3 : "translate takes exactly three arguments";
+
+        Symbol from = args.get(1);
+        if (!Literal.isLiteral(from, DataTypes.STRING)) {
+            return this;
+        }
+        Symbol to = args.get(2);
+        if (!Literal.isLiteral(to, DataTypes.STRING)) {
+            return this;
+        }
+        String fromStr = ((Input<String>) from).value();
+        String toStr = ((Input<String>) to).value();
+        return new TranslateFunction(signature, boundSignature, computeMap(fromStr, toStr));
+    }
+
+    @Override
+    public String evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>[] args) {
+        assert args.length == 3 : "translate takes exactly three arguments";
+        String source = args[0].value();
+        String from = args[1].value();
+        String to = args[2].value();
+        if (source == null || from == null || to == null) {
+            return null;
+        }
+        if (source.length() == 0 || from.length() == 0) {
+            return source;
+        }
+        Map<Character, Character> tmap = this.tmap;
+        if (tmap == null) {
+            tmap = computeMap(from, to);
+        }
+        return translate(tmap, source);
+    }
+
+    private static String translate(Map<Character, Character> tmap, String source) {
+        int sourceLen = source.length();
+        char[] result = new char[sourceLen];
+        int resultCount = 0;
+        for (int i = 0; i < sourceLen; i++) {
+            char c = source.charAt(i);
+            if (tmap.containsKey(c)) {
+                char mc = tmap.get(c);
+                if (mc != NULL) {
+                    result[resultCount++] = mc;
+                }
+            } else {
+                result[resultCount++] = c;
+            }
```

Comment:
this would avoid an additional by-key lookup which slightly improves on larger source text input according to my local benchmarks:

```suggestion
            var mc = tmap.get(c);
            if (mc == null) {
                result[resultCount++] = c;
            } else if (mc != NULL) {
                result[resultCount++] = mc;
            }
```